Update FIPS documentation for 8.x (#105041)
elastic/elasticsearch: docs/reference/security/fips-140-compliance.asciidoc

{es} offers a FIPS 140-2 compliant mode and as such can run in a FIPS 140-2
configured JVM.

IMPORTANT: The JVM bundled with {es} is not configured for FIPS 140-2. You must

> configure an external JDK and it highly recommended to use a FIPS 140-2 certified Java Security Provider.

Reviewer: I'm not sure about this change, could you please clarify more? If you don't use a certified provider, the whole setup will not be compliant.

> configure an external JDK and it highly recommended to use a FIPS 140-2 certified Java Security Provider.

Author: ```suggestion
configure an external JDK with a FIPS 140-2 certified Java Security Provider.
```

Refer to the {es}
https://www.elastic.co/support/matrix#matrix_jvm[JVM support matrix] for
supported JVM configurations. See https://www.elastic.co/subscriptions[subscriptions] for required licensing.

> Compliance with FIPS 140-2 is achieved by using only FIPS approved / NIST recommended cryptographic algorithms. This can be achieved by the following:

Reviewer 2: I'm not sure we want to make definitive statements about to achieve "compliance". 
For example, what about the Operating System? Container runtime? Hardware crypto devices?  

I'd be more comfortable stating either:
1. A negative "You are not in compliance unless... "
2. A statement that appears less exhaustive "Compliance requires that ..."

The old wording was problematic as well, but at least it was prefixed with `For {es}, ...`

Author: ++, how is this : https://github.com/elastic/elasticsearch/pull/105041/commits/4b1855f6d7a2d299125693eeb7fb38a0499f630a ?

Reviewer 2: LGTM

> Compliance with FIPS 140-2 is achieved by using only FIPS approved / NIST recommended cryptographic algorithms. This can be achieved by the following:

Reviewer: This may be confusing and read as this is the only required step to achieve compliance. I think that it's important to mention that you also need the compliant JDK.

- Installation and configuration of a FIPS certified Java security provider.
- Ensuring the configuration of {es} is FIPS 140-2 compliant as documented below.
- Setting `xpack.security.fips_mode.enabled` to `true` in `elasticsearch.yml`. Note - this setting alone is not sufficient to be compliant
with FIPS 140-2.

[discrete]
=== Configuring {es} for FIPS 140-2

Detailed instructions for the configuration required for FIPS 140-2 compliance is beyond the scope of this document.  It is the responsibility
of the user to ensure compliance with FIPS 140-2. {es} has been tested with a specific configuration described below. However, there are
other configurations possible to achieve compliance.

> The following is a high-level overview of the steps required to configure a FIPS certified security provider:

Author: ```suggestion
The following is a high-level overview of the required configuration:
```

[discrete]
[[java-security-provider]]
==== Java security provider

Detailed instructions for installation and configuration of a FIPS certified Java security provider is beyond the scope of this document.
Specifically, a FIPS certified
https://docs.oracle.com/en/java/javase/17/security/java-cryptography-architecture-jca-reference-guide.html[JCA] and
https://docs.oracle.com/en/java/javase/17/security/java-secure-socket-extension-jsse-reference-guide.html[JSSE] implementation are strongly
recommended to ensure only the FIPS approved / NIST recommended cryptographic algorithms are in use.

Elasticsearch has been tested with Bouncy Castle's https://repo1.maven.org/maven2/org/bouncycastle/bc-fips/1.0.2.4/bc-fips-1.0.2.4.jar[bc-fips 1.0.2.4]
and https://repo1.maven.org/maven2/org/bouncycastle/bctls-fips/1.0.17/bctls-fips-1.0.17.jar[bctls-fips 1.0.17] but other
FIPS certified providers may be used. Elasticsearch does not ship with a FIPS certified provider. It is the responsibility of the user
to install and configure the security provider to ensure compliance with FIPS 140-2. Using a FIPS certified provider will ensure that only
approved cryptographic algorithms are used.

> To configure {es} to use additional security provider(s) configure {es}'s JVM property `java.security.properties` to point to a file

Reviewer 2: It would be good to link to the instructions on configuring a JVM property
```suggestion
To configure {es} to use additional security provider(s) configure {es}'s <<set-jvm-options, JVM property>> `java.security.properties` to point to a file
```

> (https://raw.githubusercontent.com/elastic/elasticsearch/main/build-tools-internal/src/main/resources/fips_java.security[example]) in {es}'s

Reviewer 2: I'm hesitant to link to our test policy without additional context. I know we have some disclaimers above, but I don't think it will make much difference - people will copy our configuration and expect us to take responsibility for it.

I think we ought to tell them to review the required configuration for their security provider.

Author: >  people will copy our configuration and expect us to take responsibility for it.

I think that is worth the risk, we can remove it if it becomes a problem but there is no way someone can figure out the magic from scratch. I banged my head against my desk for a couple days just trying to figure out the few new requirements for the updated BC jars.

`config` directory. Ensure the FIPS certified security provider is configured with the lowest order. This file should contain the necessary
configuration to instruct Java to use the FIPS certified security provider.

[discrete]
[[java-security-manager]]
==== Java security manager

All code running in {es} is subject to the security restrictions enforced by the Java security manager.
The Java security manager requires explicit grants permissions to perform low level Java operations. The Java security provider is not exempt from the

> restrictions enforced by the security manager and as such it may require explicit grants to perform it's low level Java operations.

Reviewer 2: ```suggestion
restrictions enforced by the security manager and as such it may require explicit grants to perform its low level Java operations.
```

> All code running in {es} is subject to the security restrictions enforced by the Java security manager.
> The Java security manager requires explicit grants permissions to perform low level Java operations. The Java security provider is not exempt from the
> restrictions enforced by the security manager and as such it may require explicit grants to perform it's low level Java operations.

Reviewer 2: I felt like this section went into the wrong set of details. I suspect most customers don't care about what the security manager does, but do care about what they have to do to make stuff work.
 ```suggestion
 
All code running in {es} is subject to the security restrictions enforced by the Java security manager.
The security provider you have installed and configured may require additional permissions in order to function correctly. You can grant these permissions by providing your own 
https://docs.oracle.com/javase/8/docs/technotes/guides/security/PolicyFiles.html#FileSyntax[Java security policy]
```

Author: SGTM: couldn't commit the suggestion directly, so copy/pasted here: https://github.com/elastic/elasticsearch/pull/105041/commits/04b6e3b5cd934e2d3e77556fc7c535dbf4fc1df1

> All code running in {es} is subject to the security restrictions enforced by the Java security manager.
> The Java security manager requires explicit grants permissions to perform low level Java operations. The Java security provider is not exempt from the
> restrictions enforced by the security manager and as such it may require explicit grants to perform its low level Java operations.

Author: ```suggestion
All code running in {es} is subject to the security restrictions enforced by the Java security manager.
The security provider you have installed and configured may require additional permissions in order to function correctly. You can grant these permissions by providing your own 
https://docs.oracle.com/javase/8/docs/technotes/guides/security/PolicyFiles.html#FileSyntax[Java security policy]
```

Users must ensure that the security settings are configured correctly for FIPS-140-2 compliance. This is only applicable for new installations.

* Optional: Set `xpack.security.fips_mode.required_providers` in `elasticsearch.yml` to ensure the required security providers (8.13+).
see xref:verify-security-provider[below].

[discrete]
[[verify-security-provider]]
==== Verify the security provider is installed

To verify that the security provider is installed and in use, you can use any of the following steps:

* Verify the required security providers are configured with the lowest order in the file pointed to by `java.security.properties`.
For example, `security.provider.1` is a lower order than `security.provider.2`

* Set `xpack.security.fips_mode.required_providers` in `elasticsearch.yml` to the list of required security providers.
This setting is used to ensure that the correct security provider is installed and configured. (8.13+)
If the security provider is not installed correctly, {es} will fail to start. `["BCFIPS", "BCJSSE"]` are the values to
use for Bouncy Castle's FIPS JCE and JSSE certified provider.

> 

Reviewer 2: ```suggestion

[source,yaml]
--------------------------------------------------
xpack.security.fips_mode.enabled: true
xpack.security.autoconfiguration.enabled: false
xpack.security.fips_mode.required_providers: ["BCFIPS", "BCJSSE"]
--------------------------------------------------
```

ones will be created using the algorithm you have selected.

[discrete]
[[configuring-es-yml]]

> ==== Verify the security provider is installed

Author: ```suggestion
==== Configure {es} elasticsearch.yml
```
